library/core/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,60 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
```

Comment:
Is the intent to commit to this property for all $X for all time, including ones added in the future? Or does $X range over a fixed set of ABI's that are established todya.

Reply:
This was intended to apply universally.

Reply:
I can conceptualize a pair of ABIs existing that are the same, except that one doesn't have a concept of unwinding, and that the other reserves a clobbered register used for argument passing to be an exception pointer indicating unwind. Or any scheme implementing truly cooperative "inline" unwinding instead of sidechannel/externally driven unwinding.

We can commit to calling any such pairs differently than `"$X"` and `"$X-unwind"`, such that those pairs always solely differ by whether unwinds are permitted via their reserved sidechannel, so long as we do so knowingly.

It wouldn't keep me from checking an FCP box — I don't mind making that commitment — but I'd have a very slight preference to weaken this to "generally" and saying to check the actual used ABIs for compatibility.

(Are function signatures considered compatible in C++ if they only differ by `nounwind` condition? Or is that considered UB or IDB at best?)

Reply:
>  check the actual used ABIs for compatibility.

Check them where though?
Or should this list the values of `$X` for this this is definitely the case? But if we say this works for "C-unwind" then we are saying no platform default "C" ABI will use such an ABI you describe.

AFAIK on the LLVM side, "nounwind" is considered an implementation hint. I would expect it to be legal for a pass to drop the attribute.

Reply:
In other words, even if we are just promising that C and C-unwind are compatible, we're already making claims about future targets and their C ABI. Is that too much?

We can of course say the ABI strings must match exactly. I'm totally fine with that, I was just worried it might be too restrictive.

Reply:
> Or any scheme implementing truly cooperative "inline" unwinding instead of sidechannel/externally driven unwinding.

I thought about that too, but I think that would need to be a completely separate ABI, not an `-unwind` variant of an ABI that otherwise has no unwinding support. The ABI strings without `-unwind` (and excluding `Rust`) are our attempts to match existing non-Rust ABIs; the `-unwind` part is something we invented which only makes sense if the ABI itself already supports unwinding. (I.e. the only reason `C-unwind` makes sense is because standard implementations of the C ABI permit C++ style exception-propagation; and in general, the only reason every non-Rust ABI string has an `-unwind` variant is because every existing common ABI definition permits C++ style unwinding.) This is why `-unwind` is specifically defined to imply the "platform-native" unwind mechanism, _not_ something that Rust can control. (The `Rust` ABI is unique in this regard: the `-unwind` RFC specifically leaves space for the `Rust` ABI to have a different unwinding mechanism in the future.)

An ABI that has a completely separate mechanism for exception handling would, I think, not be something that users could opt out of by using `C`; it would presumably not be compatible with `C`. For instance, I believe there was an ISO proposal to formally specify a shared error-propagation mechanism that was effectively cooperative-unwinding; if that got stabilized, and if we wanted to represent that with an ABI string, it would be different from `C-unwind`.

> [An ABI could] reserve a clobbered register used for argument passing to be an exception pointer indicating unwind

That concern seems more likely to me, although I'm still not sure that would really count as the "same ABI" modulo unwinding. The `-unwind` ABIs are essentially a way for Rust users to treat non-Rust ABIs _as if_ they didn't support unwinding; they don't actually imply that external functions with those ABIs _can't_ unwind.

Reply:
@CAD97 how do you propose we proceed here?

Reply:
Following [discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/ABI.20compaibility.20docs/near/395554002) I have now restricted this to the unwind ABIs that we currently have:
```
/// The ABI strings are guaranteed to be compatible if they are the same, or if the caller ABI
/// string is `$X-unwind` and the callee ABI string is `$X`, where `$X` is one of the following:
/// "C", "aapcs", "fastcall", "stdcall", "system", "sysv64", "thiscall", "vectorcall", "win64".
```
I think this should resolve this concern.